micrometer-java21/src/main/java/io/micrometer/java21/instrument/binder/jfr/JfrVirtualThreadEventMetrics.java

```diff
@@ -0,0 +1,113 @@
+ * limitations under the License.
+ */
+package io.micrometer.java21.instrument.binder.jfr;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.core.instrument.binder.MeterBinder;
+import jdk.jfr.consumer.RecordingStream;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Objects;
+
+import static java.util.Collections.emptyList;
+
+public class JfrVirtualThreadEventMetrics implements MeterBinder, Closeable {
+
+    private static final String PINNED_EVENT = "jdk.VirtualThreadPinned";
+
+    private static final String SUBMIT_FAILED_EVENT = "jdk.VirtualThreadSubmitFailed";
+
+    private final RecordingSettings settings;
+
+    private final Iterable<Tag> tags;
+
+    private boolean started = false;
+
+    private RecordingStream recordingStream;
+
+    public JfrVirtualThreadEventMetrics() {
+        this(new RecordingSettings(), emptyList());
+    }
+
+    public JfrVirtualThreadEventMetrics(Iterable<Tag> tags) {
+        this(new RecordingSettings(), tags);
+    }
+
+    public JfrVirtualThreadEventMetrics(RecordingSettings settings, Iterable<Tag> tags) {
+        this.settings = settings;
+        this.tags = tags;
+    }
+
+    @Override
+    public void bindTo(MeterRegistry registry) {
+        if (started) {
+            return;
+        }
+
+        started = true;
+        recordingStream = createRecordingStream(settings);
+
+        final Timer pinnedTimer = Timer.builder("jvm.virtual.thread.pinned")
+            .tags(tags)
+            .description("The duration of virtual threads that were pinned to a physical thread")
+            .register(registry);
+
+        final Counter submitFailedCounter = Counter.builder("jvm.virtual.thread.submit.failed")
+            .tags(tags)
+            .description("The number of virtual thread submissions that failed")
+            .register(registry);
+
+        recordingStream.onEvent(PINNED_EVENT, event -> pinnedTimer.record(event.getDuration()));
+        recordingStream.onEvent(SUBMIT_FAILED_EVENT, event -> submitFailedCounter.increment());
+    }
+
+    protected RecordingStream createRecordingStream(RecordingSettings settings) {
+        final RecordingStream recordingStream = new RecordingStream();
+        recordingStream.enable(PINNED_EVENT).withThreshold(settings.pinnedThreshold);
+        recordingStream.enable(SUBMIT_FAILED_EVENT);
+        recordingStream.setMaxAge(settings.maxAge);
+        recordingStream.setMaxSize(settings.maxSizeBytes);
+        recordingStream.startAsync();
+        return recordingStream;
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (started) {
+            recordingStream.close();
+        }
+    }
+
+    public record RecordingSettings(Duration maxAge, long maxSizeBytes, Duration pinnedThreshold) {
```

Comment:
We usually call these config, i.e. `RecordingConfig`, I would make this private in the first round and open up as necessary.